doc/design/program.md

```diff
@@ -2,7 +2,7 @@
 
 ## Compile and Execution
 
-A PaddlePaddle program consists of two parts -- the first generates a `ProgramDesc` protobuf message that describes the program, and the second runs this message using a C++ class `Executor`.
+A PaddlePaddle program consists of three parts -- the first generates a `ProgramDesc` protobuf message that describes the program, the second optimizes this message using a C++ class `Optimizer` and generates an `ExecutionPlan` protobuf messages, and the third run the message using a C++ class `Executor`.
```

Comment:
- ExecutionPlan is not dependent on optimizer. In an inference ProgramDesc, we can also have a ExecutionPlan.
- In which time we can decide the device where an operator runs. In current code, an operator has CPU kernel and GPU kernel. At running time, the kernel is decided by input tensor's place. Actually, it's decided in running time.
Since we have ExecutionPlace which is a proto message storing device information, the device must  be decided iat compile time.

In a word, we still have two parts, compile-time and run-time. At compile-time, we will generate two proto message, the first is ProgramDesc and the second is ExecutionPlan. 
The ExecutionPlan is set by users' configuration and Paddle's own auto device placement policy. If user switch to another hardware environment, and he/she do not want to provide a ExecutionPlan, Paddle can generate a ExecutionPlan under Paddle's own auto device placement policy. 

A interface could be:
```
void GenerateExecutionPlan(const ProgramDesc& input, OpDeviceMap* user_config, ExecutionPlan* output);
```
The user_config could be null.

Reply:
> ExecutionPlan is not dependent on optimizer. In an inference ProgramDesc, we can also have a ExecutionPlan.

Agree, will find a better name for `optimizer`.

> the kernel is decided by the place of DeviceContext. Actually, it's decided in running time.

Understand, but I think deciding at runtime make us no way to control where to place the OP. Being able to control it is very important.

> In a word, we still have two parts, compile-time and run-time. At compile-time, we will generate two proto message, the first is ProgramDesc and the second is ExecutionPlan.

Agree.

> An interface could be: `void GenerateExecutionPlan(const ProgramDesc& input, OpDeviceMap* user_config, ExecutionPlan* output);`

The `user_config` should be part of `ProgramDesc`, since `ProgramDesc` describes what the user wants.